In [1]:
# Initialize OK
from client.api.notebook import Notebook
ok = Notebook('proj1a.ok')

Assignment: proj1a
OK, version v1.13.11



# Project 1A: Food Safety
## Cleaning and Exploring Data with Pandas
## Due Date: Monday 02/17, 11:59 PM
## Collaboration Policy

Data science is a collaborative activity. While you may talk with others about
the project, we ask that you **write your solutions individually**. If you do
discuss the assignments with others please **include their names** at the top
of your notebook.

**Collaborators**: *list collaborators here*


## This Assignment

In this project, we will investigate restaurant food safety scores for restaurants in San Francisco. The scores and violation information have been [made available by the San Francisco Department of Public Health](https://data.sfgov.org/Health-and-Social-Services/Restaurant-Scores-LIVES-Standard/pyih-qa8i). The main goal for this assignment is to walk through the process of Data Cleaning and EDA. 

We have broken this two week project into two parts.  The first part (week 1) will focus on basic data cleaning and analysis using Pandas.  The second part (week 2) will dive into deeper analysis and data visualization.

As we clean and explore these data, you will gain practice with:
* Reading simple csv files and using Pandas
* Working with data at different levels of granularity
* Identifying the type of data collected, missing values, anomalies, etc.
* Exploring characteristics and distributions of individual variables

## Score Breakdown 
Question | Points
--- | ---
1a | 3
1b | 2
1c | 1
1d | 1
1e | 1
2a | 1
2b | 1
2ci | 1
2cii | 1
2d | 2
3a | 2
3b | 2
3ci | 1
3cii | 1
3d | 3
3e | 2
3f | 2
4a | 1
4bi | 1
4bii | 1
4biii | 1
4ci | 1
4cii | 1
4ciii | 1
4civ | 1
4di | 2
4dii | 2
4e | 2
5a | 2
5b | 3
5c | 2
6 | 3
Total | 51

## Before You Start

For each question in the assignment, please write down your answer in the answer cell(s) right below the question. 

We understand that it is helpful to have extra cells breaking down the process towards reaching your final answer. If you happen to create new cells below your answer to run codes, **NEVER** add cells between a question cell and the answer cell below it. It will cause errors when we run the autograder, and it will sometimes cause a failure to generate the PDF file.

**Important note: The local autograder tests will not be comprehensive. You can pass the automated tests in your notebook but still fail tests in the autograder.** Please be sure to check your results carefully.

Finally, unless we state otherwise, try to avoid using python for loops or list comprehensions.  The majority of this assignment can be done using builtin commands in Pandas and numpy.  


In [2]:
import numpy as np
import pandas as pd 

import bz2 # Used to read compressed data
import os # Used to interact with the file system

## Obtaining the Data

### File Systems and I/O



In general, we will focus on using python commands to investigate files.  However, it can sometimes be easier to use shell commands in your local operating system.  The following cells demonstrate how to do this.


The command below starts with `!`. This tells our Jupyter notebook to pass this command to the operating system. In this case, the command is the `ls` POSIX command which lists all files in the current directory. Note what `!ls data` outputs.  

*Note this `!` only works in ipython shells (Jupyter Notebooks). And the `ls` (list) command only works in posix environments and may not work on default Windows systems.*

In [3]:
!ls 

data  proj1a.ipynb  proj1a.ok  tests


In [4]:
!ls data

bus.csv.bz2  ins2vio.csv.bz2  ins.csv.bz2  sf_zipcodes.json  vio.csv.bz2


We are going to use the [`pathlib`](https://docs.python.org/3/library/pathlib.html#basic-use) module to represent our file system paths and perform operations which allow us to learn more about the contents of our data. Note what `pathlib.Path.cwd()` outputs in relation to the output of `!ls` above.

<br/><br/><br/>

---

<br/><br/><br/>

# 1: Examining the Files
Let's first focus on understanding the structure of the data; this involves answering questions such as:

* How much data do we have?
* Is the data in a standard format or encoding?
* Is the data organized in records?
* What are the fields in each record?

Let's start by looking at the contents of `data/`. This is not just a single file but rather a directory of multiple compressed files. We could inspect our data by uncompressing each file but in this project we're going to do almost everything in Python for maximum portability.


You will want to use a few useful python functions.  To move through the local filesystem you can use the `Path` module in `pathlinb`.  For example, to list the current directory you can [Path.cwd](https://docs.python.org/3/library/pathlib.html#pathlib.Path.cwd).

In [5]:
from pathlib import Path

Path.cwd()

PosixPath('/home/jovyan/sp20/proj/proj1a')

The function returns a `pathlib.Path` object representing the location of the file.  It can also be used to list contents of directories and many other things. 

You will also need to work with `bzip2` files and you will want to be able to read their contents using the [bz2](https://docs.python.org/3/library/bz2.html) python library.

In [6]:
with bz2.open("data/bus.csv.bz2", "r") as f:
    print("The first line:", "\n\t", f.readline())

The first line: 
	 b'"business id column","name","address","city","state","postal_code","latitude","longitude","phone_number"\n'


<br/><br/><br/>

---

## Question 1a:

Implement the `list_files`, `get_file_size`, and `get_linecount_bz2` functions to return the list of files in the directory, the sizes (in bytes) of a file, and the number of lines in the file.  Note the last `get_linecount_bz2` should not produce any intermediate files in the filesystem and should avoid storing the entire file in memory (don't do `len(file.readlines())`).

**Hints:** *You might find the following documentation useful:*
1. [Python pathlib](https://docs.python.org/3.7/library/pathlib.html#pathlib.Path.iterdir)
1. [bz2](https://docs.python.org/3/library/bz2.html#bz2.open)


<!--
BEGIN QUESTION
name: q1a
points: 3
-->

In [7]:
def list_files(directory):
    """
    Return a list of pathlib.Path objects for the files in the directory.
    
    directory: a string describing the directory to list 
        for example 'data/'
    """
    return [p for p in Path(directory).iterdir() if p.is_file()]

def get_file_size(file_name):
    """
    Return file size for a given filename.
    """ 
    return Path(file_name).stat().st_size

def get_linecount_bz2(file_name):
    """
    Returns the number of lines in bz2 file.  
    """ 
    count=0
    for line in bz2.open(file_name):
        count += 1
    return count

In [8]:
ok.grade("q1a");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 3
    Failed: 0
[ooooooooook] 100.0% passed



Now, let's see the file size and the number of lines for each data file.  If you implemented the above code correctly the following cell should produce the following (the columns may be in a different order):

<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
       <th>linecount</th>
      <th>name</th>
      <th>size</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>2</th>
       <td>66</td>
      <td>data/vio.csv.bz2</td>
      <td>1337</td>
    </tr>
    <tr>
      <th>3</th>
      <td>26664</td>
      <td>data/ins.csv.bz2</td>
      <td>110843</td>
    </tr>
    <tr>
      <th>0</th>
      <td>6254</td>
      <td>data/bus.csv.bz2</td>
      <td>113522</td>
    </tr>
    <tr>
      <th>1</th>
      <td>40211</td>
      <td>data/ins2vio.csv.bz2</td>
      <td>146937</td>
    </tr>
  </tbody>
</table>

In [9]:
info = []
for f in list_files("data/"):
    name = str(f)
    if name[-3:] == "bz2": 
        size = get_file_size(f)
        linecount = get_linecount_bz2(f)
        info.append({"name": name, "size": size, "linecount": linecount})

file_info = pd.DataFrame(info).sort_values("size")
file_info

name    size  linecount
3      data/vio.csv.bz2    1337         66
1      data/ins.csv.bz2  110843      26664
2      data/bus.csv.bz2  113522       6254
0  data/ins2vio.csv.bz2  146937      40211

<br/><br/><br/>

---

## Question 1b: Programatically Looking Inside the Files

Implement the following function `head_bz2` to return a list of the first `nlines` lines of each file. 
Using your `head_bz2` function implement the following `print_head_bz2` function that uses `print()` to print the filename followed by the first `nlines` of each file and their line numbers in the following format.

```
data/bus.csv.bz2
0 :	 b'"business id column","name","address","city","state","postal_code","latitude","longitude","phone_number"\n'
1 :	 b'"1000","HEUNG YUEN RESTAURANT","3279 22nd St","San Francisco","CA","94110","37.755282","-122.420493","-9999"\n'
2 :	 b'"100010","ILLY CAFFE SF_PIER 39","PIER 39  K-106-B","San Francisco","CA","94133","-9999","-9999","+14154827284"\n'
3 :	 b'"100017","AMICI\'S EAST COAST PIZZERIA","475 06th St","San Francisco","CA","94103","-9999","-9999","+14155279839"\n'
4 :	 b'"100026","LOCAL CATERING","1566 CARROLL AVE","San Francisco","CA","94124","-9999","-9999","+14155860315"\n'
```

Do not read the entire file contents!

<!--
BEGIN QUESTION
name: q1b
points: 2
-->

In [10]:
def head_bz2(file, nlines=5):
    """
    Return a list of the first nlines lines of filename
    """
    f=[]
    stoper=0
    for line in bz2.open(file):
        if stoper < nlines :
            stoper +=1
            f.append(line)
        else:
            return f
    
def print_head_bz2(file, nlines=5):
    """
    Print a list of the first nlines lines of filename
    """
    stoper=0
    f=head_bz2(file, nlines)
    print(file)
    for x in f:
        if stoper < nlines :
            print(str(stoper) + ' :     ' + str(x))
            stoper+=1

In [11]:
ok.grade("q1b");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 2
    Failed: 0
[ooooooooook] 100.0% passed



The following should display the filename and head of the file for all the files in data:

In [12]:
for file in list_files("data/"):
    if str(file)[-3:] == "bz2":  
        print_head_bz2(file)
        print()

data/ins2vio.csv.bz2
0 :     b'"iid","vid"\n'
1 :     b'"97975_20190725","103124"\n'
2 :     b'"85986_20161011","103114"\n'
3 :     b'"95754_20190327","103124"\n'
4 :     b'"77005_20170429","103120"\n'

data/ins.csv.bz2
0 :     b'"iid","date","score","type"\n'
1 :     b'"100010_20190329","03/29/2019 12:00:00 AM","-1","New Construction"\n'
2 :     b'"100010_20190403","04/03/2019 12:00:00 AM","100","Routine - Unscheduled"\n'
3 :     b'"100017_20190417","04/17/2019 12:00:00 AM","-1","New Ownership"\n'
4 :     b'"100017_20190816","08/16/2019 12:00:00 AM","91","Routine - Unscheduled"\n'

data/bus.csv.bz2
0 :     b'"business id column","name","address","city","state","postal_code","latitude","longitude","phone_number"\n'
1 :     b'"1000","HEUNG YUEN RESTAURANT","3279 22nd St","San Francisco","CA","94110","37.755282","-122.420493","-9999"\n'
2 :     b'"100010","ILLY CAFFE SF_PIER 39","PIER 39  K-106-B","San Francisco","CA","94133","-9999","-9999","+14154827284"\n'
3 :     b'"100017","AMICI\'S

<br/><br/><br/>

---

## Question 1c: Thinking about the files

Answer the following questions by filling in the correct boolean values in the following variables:

1. The bus.csv.bz2 file appears to be tab delimited.
1. The values all appear to be quoted.


<!--
BEGIN QUESTION
name: q1c
points: 1
-->

In [13]:
# True or False: The bus.csv.bz2 file appears to be tab delimited.
q1c_1 = False

# True or False: The values all appear to be quoted.
q1c_2 = True

In [14]:
ok.grade("q1c");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 2
    Failed: 0
[ooooooooook] 100.0% passed



<br/><br/><br/>

---

## Question 1d: Reading in the Files

Based on the above information, let's attempt to load `bus.csv`, `ins2vio.csv`, `vio.csv` and `ins.csv` into Pandas dataframes with the following names: `bus`, `ins2vio`, `vio`, `ins` respectively.


<!--
BEGIN QUESTION
name: q1d
points: 1
-->

In [15]:
# path to directory containing data
bus = pd.read_csv('data/bus.csv.bz2')
ins = pd.read_csv('data/ins.csv.bz2')
ins2vio = pd.read_csv('data/ins2vio.csv.bz2')
vio = pd.read_csv('data/vio.csv.bz2')

In [16]:
ok.grade("q1d");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 4
    Failed: 0
[ooooooooook] 100.0% passed



Now that you've read in the files, you can try some `pd.DataFrame` methods ([docs](https://pandas.pydata.org/pandas-docs/version/0.21/generated/pandas.DataFrame.html)).
You can use the `DataFrame.head` method to show the top few lines of the `bus`, `ins`, `ins2vio` and `vio` dataframes. 

In [17]:
bus.head()

business id column                         name                 address  \
0                1000        HEUNG YUEN RESTAURANT            3279 22nd St   
1              100010        ILLY CAFFE SF_PIER 39        PIER 39  K-106-B   
2              100017  AMICI'S EAST COAST PIZZERIA             475 06th St   
3              100026               LOCAL CATERING        1566 CARROLL AVE   
4              100030             OUI OUI! MACARON  2200 JERROLD AVE STE C   

            city state postal_code     latitude    longitude  phone_number  
0  San Francisco    CA       94110    37.755282  -122.420493         -9999  
1  San Francisco    CA       94133 -9999.000000 -9999.000000   14154827284  
2  San Francisco    CA       94103 -9999.000000 -9999.000000   14155279839  
3  San Francisco    CA       94124 -9999.000000 -9999.000000   14155860315  
4  San Francisco    CA       94124 -9999.000000 -9999.000000   14159702675

In [18]:
ins.head()

iid                    date  score                   type
0  100010_20190329  03/29/2019 12:00:00 AM     -1       New Construction
1  100010_20190403  04/03/2019 12:00:00 AM    100  Routine - Unscheduled
2  100017_20190417  04/17/2019 12:00:00 AM     -1          New Ownership
3  100017_20190816  08/16/2019 12:00:00 AM     91  Routine - Unscheduled
4  100017_20190826  08/26/2019 12:00:00 AM     -1  Reinspection/Followup

In [19]:
ins2vio.head()

iid     vid
0  97975_20190725  103124
1  85986_20161011  103114
2  95754_20190327  103124
3  77005_20170429  103120
4   4794_20181030  103138

In [20]:
vio.head()

description  risk_category     vid
0  Consumer advisory not provided for raw or unde...  Moderate Risk  103128
1                   Contaminated or adulterated food      High Risk  103108
2          Discharge from employee nose mouth or eye  Moderate Risk  103117
3                         Employee eating or smoking  Moderate Risk  103118
4                             Food in poor condition  Moderate Risk  103123

<br/><br/><br/>

---

## Question 1e: Identifying Issues with the Data

Use the `head` command on your four files again. This time, describe at least one potential problem with the data you see. Consider issues with missing values and bad data.

Please write your answer in the markdown cell below. You may create new cells below your answer to run code, but **please never add cells between a question cell and the answer cell below it.**

<!--
BEGIN QUESTION
name: q1e
manual: True
points: 1
-->
<!-- EXPORT TO PDF -->

One potential problem of these data sets is negative or not realistic numbers set of data. For example, in bus, they have -9999 latitdue adn longtidue, ins have negative scores.

<br/><br/><br/>

---

<br/><br/><br/>

# 2: Examining the Business Data File

From its name alone, we expect the `bus.csv` file to contain information about the restaurants. Let's investigate the granularity of this dataset.

In [21]:
bus.head()

business id column                         name                 address  \
0                1000        HEUNG YUEN RESTAURANT            3279 22nd St   
1              100010        ILLY CAFFE SF_PIER 39        PIER 39  K-106-B   
2              100017  AMICI'S EAST COAST PIZZERIA             475 06th St   
3              100026               LOCAL CATERING        1566 CARROLL AVE   
4              100030             OUI OUI! MACARON  2200 JERROLD AVE STE C   

            city state postal_code     latitude    longitude  phone_number  
0  San Francisco    CA       94110    37.755282  -122.420493         -9999  
1  San Francisco    CA       94133 -9999.000000 -9999.000000   14154827284  
2  San Francisco    CA       94103 -9999.000000 -9999.000000   14155279839  
3  San Francisco    CA       94124 -9999.000000 -9999.000000   14155860315  
4  San Francisco    CA       94124 -9999.000000 -9999.000000   14159702675

<br/><br/><br/>

---

## Question 2a

The `bus` dataframe contains a column called `business id column` which probably corresponds to a unique business id.  However, let's first rename that column to `bid`.  Modify the `bus` dataframe by renaming that column to `bid`.

**Note**: In practice we might want to do this renaming when the table is loaded but for grading purposes we will do it here.

<!--
BEGIN QUESTION
name: q2a
points: 1
-->

In [22]:
bus = bus.rename(columns={"business id column":"bid"})

In [23]:
ok.grade("q2a");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 1
    Failed: 0
[ooooooooook] 100.0% passed



<br/><br/><br/>

---

## Question 2b

Examining the entries in `bus`, is the `bid` unique for each record (i.e. each row of data)? Your code should compute the answer, i.e. don't just hard code `True` or `False`.

Hint: use `value_counts()` or `unique()` to determine if the `bid` series has any duplicates.

<!--
BEGIN QUESTION
name: q2b
points: 1
-->

In [24]:
is_bid_unique = (len(bus.bid.unique().tolist())==len(bus.bid))

In [25]:
ok.grade("q2b");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 1
    Failed: 0
[ooooooooook] 100.0% passed



<br/><br/><br/>

---


## Question 2c

In the two cells below create two **series** 

1. where the index is the `name` of the business and the value is the number of records with that `name`
2. where the index is the `address` of the business and the value is the number of records with that `address`

Order both series in descending order by count. You may need to use `groupby()`, `size()`, `sort_values()`, or `value_counts()`. 

**Step 1**

<!--
BEGIN QUESTION
name: q2ci
points: 1
-->

In [26]:
name_counts = pd.Series(bus.name.value_counts(),index=bus.name.unique())
name_counts.head(20)

HEUNG YUEN RESTAURANT                      1
ILLY CAFFE SF_PIER 39                      1
AMICI'S EAST COAST PIZZERIA                2
LOCAL CATERING                             1
OUI OUI! MACARON                           1
Hula Truck (#2)                            1
GENKI CREPES & MINI MART                   1
UNCLE LEE CAFE                             1
Twirl and Dip                              2
SF PITA HUB                                1
DUMPLING ALLEY                             1
Mission Blue                               1
SUBWAY SANDWICHES #7307                    1
POSITIVE FOODS                             3
THE MATTERHORN RESTAURANT AND BAKERY       1
SLN CTRNG                                  1
THE EPICUREAN TRADER                       1
FRJTZ KITCHEN                              1
THE LITTLE CHIHUAHUA MEXICAN RESTAURANT    1
GANGNAM BBQ                                1
Name: name, dtype: int64

In [27]:
ok.grade("q2ci");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 2
    Failed: 0
[ooooooooook] 100.0% passed



**Step 2**

<!--
BEGIN QUESTION
name: q2cii
points: 1
-->

In [28]:
address_counts = pd.Series(bus.address.value_counts(),index=bus.address.unique())
address_counts.head(10)

3279 22nd St                     1
PIER 39  K-106-B                 1
475 06th St                      2
1566 CARROLL AVE                 2
2200 JERROLD AVE STE C           1
2 Marina Blvd                    4
330 CLEMENT ST                   1
3608 BALBOA ST                   1
335 Martin Luther King Jr. Dr    1
475 06TH ST                      8
Name: address, dtype: int64

In [29]:
ok.grade("q2cii");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 2
    Failed: 0
[ooooooooook] 100.0% passed



<br/><br/><br/>

---


## Question 2d

Based on the above calculations answer each of the following questions by filling the value in the variable.

1. What does each record represent?  
1. What is the minimal primary key?

<!--
BEGIN QUESTION
name: q2d
points: 2
-->

In [30]:
# What does each record represent?  Valid answers are:
#    "One location of a restaurant."
#    "A chain of restaurants."
#    "A city block."
q2d_part1 = "One location of a restaurant."

# What is the minimal primary key? Valid answers are:
#    "bid"
#    "bid, name"
#    "bid, name, address"
q2d_part2 = "bid"

In [31]:
ok.grade("q2d");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 2
    Failed: 0
[ooooooooook] 100.0% passed



<br/><br/><br/>

---

<br/><br/><br/>

# 3: Cleaning the Business Data Postal Codes

The business data contains postal code information that we can use to aggregate the ratings over regions of the city.  Let's examine and clean the postal code field.  The postal code (sometimes also called a ZIP code) partitions the city into regions:

<img src="https://www.usmapguide.com/wp-content/uploads/2019/03/printable-san-francisco-zip-code-map.jpg" alt="ZIP Code Map" style="width: 600px">

<br/><br/><br/>

---


## Question 3a

How many restaurants are in each ZIP code? 

In the cell below, create a **series** where the index is the postal code and the value is the number of records with that postal code in descending order of count. You may need to use `groupby()`, `size()`, or `value_counts()`. Do you notice any odd/invalid zip codes?

<!--
BEGIN QUESTION
name: q3a
points: 1
-->

In [32]:
zip_counts = pd.Series(bus.postal_code.value_counts(),index=bus.postal_code.unique())
print(zip_counts.to_string())

94110         555
94133         398
94103         562
94124         218
94123         177
94118         231
94121         157
94134          82
94114         200
94109         382
94102         456
94132         132
94116          97
-9999         194
94107         408
94105         249
94108         229
94117         189
94158          90
94112         192
94127          67
94105-1420      1
94111         259
94122         255
94115         230
94104         142
94122-1909      1
94131          49
94117-3504      1
94518           1
95105           1
94013           2
94130           8
941102019       1
941             1
941033148       1
CA              2
92672           1
94120           1
94143           5
94101           2
94014           1
94129           1
94602           1
Ca              1
94080           1
00000           1
94188           2
64110           1
94544           1
94301           2
94901           1
95117           1
95133           1
95109           1
95132     

In [33]:
ok.grade("q3a");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 3
    Failed: 0
[ooooooooook] 100.0% passed



<br/><br/><br/>

--- 

## Question 3b

Answer the following questions about the `postal_code` column in the `bus` dataframe.

1. The ZIP code column is which of the following type of data:
    1. Quantitative Continuous
    1. Quantitative Discrete
    1. Qualitative Ordinal
    1. Qualitative Nominal    
1. What Python data type is used to represent a ZIP code?

*Note*: ZIP codes and postal codes are the same thing.

Please write your answers in the variables below:

<!--
BEGIN QUESTION
name: q3b
points: 2
-->

In [34]:
# The ZIP code column is which of the following type of data:
#   "Quantitative Continuous" 
#   "Quantitative Discrete"
#   "Qualitative Ordinal"
#   "Qualitative Nominal"
q3b_part1 = "Qualitative Nominal"

# What Python data type is used to represent a ZIP code? 
#    "str"
#    "int"
#    "bool"
#    "float"
q3b_part2 = "str"


In [35]:
ok.grade("q3b");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 2
    Failed: 0
[ooooooooook] 100.0% passed



<br/><br/><br/>

--- 

## Question 3c

In question 3a we noticed a large number of potentially invalid ZIP codes (e.g., "CA").  These are likely due to data entry errors.  To get a better understanding of the potential errors in the zip codes we will:

1. Import a list of valid San Francisco ZIP codes by using `pd.read_json` to load the file `data/sf_zipcodes.json` and extract a **series** of type `str` containing the valid ZIP codes.  *Hint: set `dtype` when invoking `read_json`.*
1. Construct a DataFrame containing only the businesses which DO NOT have valid ZIP codes.  You will probably want to use the `Series.isin` function. 


**Step 1**

<!--
BEGIN QUESTION
name: q3ci
points: 1
-->

In [36]:
valid_zips = pd.read_json('data/sf_zipcodes.json',dtype=str).iloc[:,0]
valid_zips.head()

0    94102
1    94103
2    94104
3    94105
4    94107
Name: zip_codes, dtype: object

In [37]:
ok.grade("q3ci");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 2
    Failed: 0
[ooooooooook] 100.0% passed



**Step 2**

<!--
BEGIN QUESTION
name: q3cii
points: 1
-->

In [38]:
invalid_zip_bus = bus[~bus["postal_code"].isin(valid_zips)]
invalid_zip_bus.head()


bid                       name                              address  \
22   100126  Lamas Peruvian Food Truck                     Private Location   
68   100417           COMPASS ONE, LLC                      1 MARKET ST. FL   
96   100660                  TEAPENTER                       1518 IRVING ST   
109  100781          LE CAFE DU SOLEIL                      200 FILLMORE ST   
144  101084             Deli North 200  1 Warriors Way Level 300 North East   

              city state postal_code  latitude  longitude  phone_number  
22   San Francisco    CA       -9999   -9999.0    -9999.0         -9999  
68   San Francisco    CA  94105-1420   -9999.0    -9999.0   14154324000  
96   San Francisco    CA  94122-1909   -9999.0    -9999.0   14155868318  
109  San Francisco    CA  94117-3504   -9999.0    -9999.0   14155614215  
144  San Francisco    CA       94518   -9999.0    -9999.0         -9999

In [39]:
ok.grade("q3cii");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 2
    Failed: 0
[ooooooooook] 100.0% passed



<br/><br/><br/>

--- 

## Question 3d

In the previous question, many of the businesses had a common invalid postal code that was likely used to code a MISSING postal code.  Do they all share a potentially "interesting address"?

In the following cell, construct a **series** that counts the number of businesses at each `address` that have this single likely MISSING postal code value.  Order the series in descending order by count. 

After examining the output.  Answer the following question by filling in the appropriate variable. If we were to drop businesses with MISSING postal code values would a particular class of business be affected?  If you are unsure try to search the web for the most common addresses.


<!--
BEGIN QUESTION
name: q3d
points: 3
-->

In [40]:
invalid_zip_list=bus[['address','postal_code']].where(bus['postal_code']=='-9999').dropna()
missing_zip_address_count = pd.Series(invalid_zip_list['address'].value_counts(),index=invalid_zip_list['address'].unique()).sort_values(ascending=False)
missing_zip_address_count



Off The Grid                  39
Off the Grid                  10
OTG                            4
Approved Private Locations     3
Approved Locations             3
                              ..
3251 20th Ave                  1
510 Stevenson St               1
610 Long Bridge                1
855 Bush St                    1
Private Location               1
Name: address, Length: 135, dtype: int64

In [41]:
ok.grade("q3d");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 3
    Failed: 0
[ooooooooook] 100.0% passed



<br/><br/><br/>

--- 

## Question 3e

**True or False**:  *If we were to drop businesses with MISSING postal code values a particular class of business will be affected.*

<!--
BEGIN QUESTION
name: q3e
points: 2
-->

In [42]:
# True or False: 
#  If we were to drop businesses with MISSING postal code values 
#   a particular class of business be affected.
q3d_true_or_false = False

In [43]:
ok.grade("q3e");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 1
    Failed: 0
[ooooooooook] 100.0% passed



<br/><br/><br/>

--- 

## Question 3f

Examine the `invalid_zip_bus` dataframe we computed above and look at the businesses that DO NOT have the special MISSING ZIP code value.  Some of the invalid postal codes are just the full 9 digit code rather than the first 5 digits.  Create a new column named `postal5` in the original `bus` dataframe which contains only the first 5 digits of the `postal_code` column.   Finally, for any of the `postal5` ZIP code entries that were not a valid San Fransisco ZIP Code (according to `valid_zips`) set the entry to `None`.  


<!--
BEGIN QUESTION
name: q3f
points: 2
-->

In [44]:
bus['postal5'] = bus['postal_code'].str[:5]
bus['postal5'] = bus['postal5'].isin(valid_zips).replace({False:None})
bus['postal5'] = bus['postal5'].replace(True,bus['postal_code'].str[:5])

# Checking the corrected postal5 column
bus.loc[invalid_zip_bus.index, ['bid', 'name', 'postal_code', 'postal5']]




bid                            name postal_code postal5
22    100126       Lamas Peruvian Food Truck       -9999    None
68    100417                COMPASS ONE, LLC  94105-1420   94105
96    100660                       TEAPENTER  94122-1909   94122
109   100781               LE CAFE DU SOLEIL  94117-3504   94117
144   101084                  Deli North 200       94518    None
...      ...                             ...         ...     ...
6173   99369                     HOTEL BIRON  94102-5917   94102
6174   99376  Mashallah Halal Food truck Ind       -9999    None
6199   99536               FAITH SANDWICH #2  94105-2907   94105
6204   99681                         Twister       95112    None
6241   99819                  CHESTNUT DINER  94123-3106   94123

[230 rows x 4 columns]

In [45]:
ok.grade("q3f");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 4
    Failed: 0
[ooooooooook] 100.0% passed



<br/><br/><br/>

---

<br/><br/><br/>

# 4: Investigate the Inspection Data

Let's now turn to the inspection DataFrame. Earlier, we found that `ins` has 4 columns named 
`iid`, `score`, `date` and `type`.  In this section, we determine the granularity of `ins` and investigate the kinds of information provided for the inspections. 

Let's start by looking again at the first 5 rows of `ins` to see what we're working with.

In [46]:
ins.head(5)

iid                    date  score                   type
0  100010_20190329  03/29/2019 12:00:00 AM     -1       New Construction
1  100010_20190403  04/03/2019 12:00:00 AM    100  Routine - Unscheduled
2  100017_20190417  04/17/2019 12:00:00 AM     -1          New Ownership
3  100017_20190816  08/16/2019 12:00:00 AM     91  Routine - Unscheduled
4  100017_20190826  08/26/2019 12:00:00 AM     -1  Reinspection/Followup

<br/><br/><br/>

---

## Question 4a

The column `iid` probably corresponds to an inspection id.  Is it a primary key?  Write an expression (line of code) that evaluates to 'True' or 'False' based on whether all the values are unique.


<!--
BEGIN QUESTION
name: q4a
points: 1
-->

In [47]:
is_ins_iid_a_primary_key = True

In [48]:
ok.grade("q4a");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 1
    Failed: 0
[ooooooooook] 100.0% passed



<br/><br/><br/>

---

## Question 4b

The column `iid` appears to be the composition of two numbers and the first number looks like a business id.  

**Part 1.**: Create a new column called `bid` in the `ins` dataframe containing just the business id.  You will want to use `ins['iid'].str` operations to do this.  Also be sure to convert the type of this column to `int`

**Part 2.**: Then compute how many values in this new column are also valid business ids (appear in the `bus['bid']` column). This is verifying a foreign key relationship. Consider using the `pd.Series.isin` function.

**Part 3.**: Answer True or False, `ins['bid']` is a foreign key reference to `bus['bid']`.


**No python `for` loops or list comprehensions required!**

**Part 1**

<!--
BEGIN QUESTION
name: q4bi
points: 1
-->

In [49]:
ins['bid']=ins['iid'].str.split('_').str[0].astype(int)

In [50]:
ok.grade("q4bi");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 3
    Failed: 0
[ooooooooook] 100.0% passed



**Part 2**

<!--
BEGIN QUESTION
name: q4bii
points: 1
-->

In [51]:
invalid_bid_count = (pd.Series(ins['bid']).isin(bus['bid'])==True).count()

In [52]:
ok.grade("q4bii");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 1
    Failed: 0
[ooooooooook] 100.0% passed



**Part 3**

<!--
BEGIN QUESTION
name: q4biii
points: 1
-->

In [53]:
# True or False: The column ins['bid'] is a foreign key 
#   referencing the bus['bid'] primary key.

q4b_is_foreign_key = True

In [54]:
ok.grade("q4biii");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 1
    Failed: 0
[ooooooooook] 100.0% passed



<br/><br/><br/>

---

## Question 4c

What if we are interested in a time component of the inspection data?  We need to examine the date column of each inspection. 

**Part 1:** What is the type of the individual `ins['date']` entries. You may want to grab the very first entry and use the `type` function in python. 

**Part 2:** Use `pd.to_datetime` to create a new `ins['timestamp']` column containing of `pd.Timestamp` objects.  These will allow us to do more date manipulation.

**Part 3:** What are the earliest and latest dates in our inspection data?  *Hint: you can use `min` and `max` on dates of the correct type.*

**Part 4:** We probably want to examine the inspections by year. Create an additional `ins['year']` column containing just the year of the inspection.  Consider using `pd.Series.dt.year` to do this.

**No python `for` loops or list comprehensions required!**

**Part 1**

<!--
BEGIN QUESTION
name: q4ci
points: 1
-->

In [55]:
ins_date_type = type(ins['date'])
ins_date_type

pandas.core.series.Series

In [56]:
ok.grade("q4ci");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 1
    Failed: 0
[ooooooooook] 100.0% passed



**Part 2**

<!--
BEGIN QUESTION
name: q4cii
points: 1
-->

In [57]:
ins['timestamp']=pd.to_datetime(ins['date'])

In [58]:
ok.grade("q4cii");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 1
    Failed: 0
[ooooooooook] 100.0% passed



**Part 3**

<!--
BEGIN QUESTION
name: q4ciii
points: 1
-->

In [59]:
earliest_date = ins['timestamp'].min()
latest_date = ins['timestamp'].max()

print("Earliest Date:", earliest_date)
print("Latest Date:", latest_date)

Earliest Date: 2016-10-04 00:00:00
Latest Date: 2019-11-28 00:00:00


In [60]:
ok.grade("q4ciii");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 2
    Failed: 0
[ooooooooook] 100.0% passed



**Part 4**

<!--
BEGIN QUESTION
name: q4civ
points: 1
-->

In [61]:
ins['year']=ins['timestamp'].dt.year

In [62]:
ok.grade("q4civ");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 1
    Failed: 0
[ooooooooook] 100.0% passed



In [63]:
ins.head()

iid                    date  score                   type  \
0  100010_20190329  03/29/2019 12:00:00 AM     -1       New Construction   
1  100010_20190403  04/03/2019 12:00:00 AM    100  Routine - Unscheduled   
2  100017_20190417  04/17/2019 12:00:00 AM     -1          New Ownership   
3  100017_20190816  08/16/2019 12:00:00 AM     91  Routine - Unscheduled   
4  100017_20190826  08/26/2019 12:00:00 AM     -1  Reinspection/Followup   

      bid  timestamp  year  
0  100010 2019-03-29  2019  
1  100010 2019-04-03  2019  
2  100017 2019-04-17  2019  
3  100017 2019-08-16  2019  
4  100017 2019-08-26  2019

<br/><br/><br/>

---

## Question 4d

What is the relationship between the type of inspection over the 2016 to 2019 timeframe? 

**Part 1**

Construct the following table by
1. Using the `pivot_table` containing the number (`size`) of inspections for the given `type` and `year`.
1. Adding an extra `Total` column to the result using `sum`
1. Sort the results in descending order by the `Total`.

<table border="1" class="dataframe">  <thead>    <tr style="text-align: right;">      <th>year</th>      <th>2016</th>      <th>2017</th>      <th>2018</th>      <th>2019</th>      <th>Total</th>    </tr>    <tr>      <th>type</th>      <th></th>      <th></th>      <th></th>      <th></th>      <th></th>    </tr>  </thead>  <tbody>    <tr>      <th>Routine - Unscheduled</th>      <td>966</td>      <td>4057</td>      <td>4373</td>      <td>4681</td>      <td>14077</td>    </tr>    <tr>      <th>Reinspection/Followup</th>      <td>445</td>      <td>1767</td>      <td>1935</td>      <td>2292</td>      <td>6439</td>    </tr>    <tr>      <th>New Ownership</th>      <td>99</td>      <td>506</td>      <td>528</td>      <td>459</td>      <td>1592</td>    </tr>    <tr>      <th>Complaint</th>      <td>91</td>      <td>418</td>      <td>512</td>      <td>437</td>      <td>1458</td>    </tr>    <tr>      <th>New Construction</th>      <td>102</td>      <td>485</td>      <td>218</td>      <td>189</td>      <td>994</td>    </tr>    <tr>      <th>Non-inspection site visit</th>      <td>51</td>      <td>276</td>      <td>253</td>      <td>231</td>      <td>811</td>    </tr>    <tr>      <th>New Ownership - Followup</th>      <td>0</td>      <td>45</td>      <td>219</td>      <td>235</td>      <td>499</td>    </tr>    <tr>      <th>Structural Inspection</th>      <td>1</td>      <td>153</td>      <td>50</td>      <td>190</td>      <td>394</td>    </tr>    <tr>      <th>Complaint Reinspection/Followup</th>      <td>19</td>      <td>68</td>      <td>70</td>      <td>70</td>      <td>227</td>    </tr>    <tr>      <th>Foodborne Illness Investigation</th>      <td>1</td>      <td>29</td>      <td>50</td>      <td>35</td>      <td>115</td>    </tr>    <tr>      <th>Routine - Scheduled</th>      <td>0</td>      <td>9</td>      <td>8</td>      <td>29</td>      <td>46</td>    </tr>    <tr>      <th>Administrative or Document Review</th>      <td>2</td>      <td>1</td>      <td>1</td>      <td>0</td>      <td>4</td>    </tr>    <tr>      <th>Multi-agency Investigation</th>      <td>0</td>      <td>0</td>      <td>1</td>      <td>2</td>      <td>3</td>    </tr>    <tr>      <th>Special Event</th>      <td>0</td>      <td>3</td>      <td>0</td>      <td>0</td>      <td>3</td>    </tr>    <tr>      <th>Community Health Assessment</th>      <td>1</td>      <td>0</td>      <td>0</td>      <td>0</td>      <td>1</td>    </tr>  </tbody></table>


**Part 2**

Based on the above analysis, which year appears to have had a lot of businesses in new buildings?

**No python `for` loops or list comprehensions required!**

**Part 1**

<!--
BEGIN QUESTION
name: q4di
points: 2
-->

In [64]:
ins_pivot = pd.pivot_table(ins,values='timestamp',index='type',columns='year',aggfunc=np.size,fill_value=0,margins=True,margins_name="Total")
ins_pivot_sorted=ins_pivot.drop(ins_pivot.index[-1]).sort_values('Total',ascending=0)
ins_pivot_sorted

year                               2016  2017  2018  2019  Total
type                                                            
Routine - Unscheduled               966  4057  4373  4681  14077
Reinspection/Followup               445  1767  1935  2292   6439
New Ownership                        99   506   528   459   1592
Complaint                            91   418   512   437   1458
New Construction                    102   485   218   189    994
Non-inspection site visit            51   276   253   231    811
New Ownership - Followup              0    45   219   235    499
Structural Inspection                 1   153    50   190    394
Complaint Reinspection/Followup      19    68    70    70    227
Foodborne Illness Investigation       1    29    50    35    115
Routine - Scheduled                   0     9     8    29     46
Administrative or Document Review     2     1     1     0      4
Multi-agency Investigation            0     0     1     2      3
Special Event                         0     3     0     0      3
Community Health Assessment           1     0     0     0      1

In [65]:
ok.grade("q4di");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 2
    Failed: 0
[ooooooooook] 100.0% passed



**Part 2**

<!--
BEGIN QUESTION
name: q4dii
points: 2
-->

In [66]:
year_of_new_construction = 2017

In [67]:
ok.grade("q4dii");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 1
    Failed: 0
[ooooooooook] 100.0% passed



<br/><br/><br/>

---

## Question 4e

Let's examine the inspection scores `ins['score']`



In [68]:
ins['score'].value_counts().head()

-1      12632
 100     1993
 96      1681
 92      1260
 94      1250
Name: score, dtype: int64

There are a large number of inspections with the `'score'` of `-1`.   These are probably missing values.  Let's see what type of inspections have scores and which do not.  Create the following dataframe using steps similar to the previous question.

You should observe that inspection scores appear only to be assigned to `Routine - Unscheduled` inspections.


<table border="1" class="dataframe">  <thead>    <tr style="text-align: right;">      <th>Missing Score</th>      <th>False</th>      <th>True</th>      <th>Total</th>    </tr>    <tr>      <th>type</th>      <th></th>      <th></th>      <th></th>    </tr>  </thead>  <tbody>    <tr>      <th>Routine - Unscheduled</th>      <td>14031</td>      <td>46</td>      <td>14077</td>    </tr>    <tr>      <th>Reinspection/Followup</th>      <td>0</td>      <td>6439</td>      <td>6439</td>    </tr>    <tr>      <th>New Ownership</th>      <td>0</td>      <td>1592</td>      <td>1592</td>    </tr>    <tr>      <th>Complaint</th>      <td>0</td>      <td>1458</td>      <td>1458</td>    </tr>    <tr>      <th>New Construction</th>      <td>0</td>      <td>994</td>      <td>994</td>    </tr>    <tr>      <th>Non-inspection site visit</th>      <td>0</td>      <td>811</td>      <td>811</td>    </tr>    <tr>      <th>New Ownership - Followup</th>      <td>0</td>      <td>499</td>      <td>499</td>    </tr>    <tr>      <th>Structural Inspection</th>      <td>0</td>      <td>394</td>      <td>394</td>    </tr>    <tr>      <th>Complaint Reinspection/Followup</th>      <td>0</td>      <td>227</td>      <td>227</td>    </tr>    <tr>      <th>Foodborne Illness Investigation</th>      <td>0</td>      <td>115</td>      <td>115</td>    </tr>    <tr>      <th>Routine - Scheduled</th>      <td>0</td>      <td>46</td>      <td>46</td>    </tr>    <tr>      <th>Administrative or Document Review</th>      <td>0</td>      <td>4</td>      <td>4</td>    </tr>    <tr>      <th>Multi-agency Investigation</th>      <td>0</td>      <td>3</td>      <td>3</td>    </tr>    <tr>      <th>Special Event</th>      <td>0</td>      <td>3</td>      <td>3</td>    </tr>    <tr>      <th>Community Health Assessment</th>      <td>0</td>      <td>1</td>      <td>1</td>    </tr>  </tbody></table>



<!--
BEGIN QUESTION
name: q4e
points: 2
-->

In [69]:
ins['Missing Score']=ins['score'].isin([-1])
ins_missing_score_pivot = pd.pivot_table(ins,index='type',values= 'score',columns='Missing Score',aggfunc="count",fill_value=0,margins=True,margins_name="Total")
ins_missing_score_pivot =ins_missing_score_pivot.drop(ins_missing_score_pivot .index[-1]).sort_values('Total',ascending=0)
ins_missing_score_pivot 

Missing Score                      False  True  Total
type                                                 
Routine - Unscheduled              14031    46  14077
Reinspection/Followup                  0  6439   6439
New Ownership                          0  1592   1592
Complaint                              0  1458   1458
New Construction                       0   994    994
Non-inspection site visit              0   811    811
New Ownership - Followup               0   499    499
Structural Inspection                  0   394    394
Complaint Reinspection/Followup        0   227    227
Foodborne Illness Investigation        0   115    115
Routine - Scheduled                    0    46     46
Administrative or Document Review      0     4      4
Multi-agency Investigation             0     3      3
Special Event                          0     3      3
Community Health Assessment            0     1      1

In [70]:
ok.grade("q4e");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 1
    Failed: 0
[ooooooooook] 100.0% passed



Notice that inspection scores appear only to be assigned to `Routine - Unscheduled` inspections. It is reasonable that for inspection types such as `New Ownership` and `Complaint` to have no associated inspection scores, but we might be curious why there are no inspection scores for the `Reinspection/Followup` inspection type.

<br/><br/><br/>

---

<br/><br/><br/>

# 5: Joining Data Across Tables

In this question we will start to connect data across mulitple tables.  We will be using the `merge` function. 

<br/><br/><br/>

--- 

## Question 5a

Let's figure out which restaurants had the lowest scores. Let's start by creating a new dataframe called `ins_named`. It should be exactly the same as `ins`, except that it should have the name and address of every business, as determined by the `bus` dataframe. 

*Hint*: Use the merge method to join the `ins` dataframe with the appropriate portion of the `bus` dataframe. See the official [documentation](https://pandas.pydata.org/pandas-docs/stable/user_guide/merging.html) on how to use `merge`.

*Note*: For quick reference, a pandas 'left' join keeps the keys from the left frame, so if ins is the left frame, all the keys from ins are kept and if a set of these keys don't have matches in the other frame, the columns from the other frame for these "unmatched" key rows contains NaNs.

<!--
BEGIN QUESTION
name: q5a
points: 1
-->

In [71]:
ins_named = pd.merge(ins,bus[['bid','name','address']],on='bid')
ins_named.head()


iid                    date  score                   type  \
0  100010_20190329  03/29/2019 12:00:00 AM     -1       New Construction   
1  100010_20190403  04/03/2019 12:00:00 AM    100  Routine - Unscheduled   
2  100017_20190417  04/17/2019 12:00:00 AM     -1          New Ownership   
3  100017_20190816  08/16/2019 12:00:00 AM     91  Routine - Unscheduled   
4  100017_20190826  08/26/2019 12:00:00 AM     -1  Reinspection/Followup   

      bid  timestamp  year  Missing Score                         name  \
0  100010 2019-03-29  2019           True        ILLY CAFFE SF_PIER 39   
1  100010 2019-04-03  2019          False        ILLY CAFFE SF_PIER 39   
2  100017 2019-04-17  2019           True  AMICI'S EAST COAST PIZZERIA   
3  100017 2019-08-16  2019          False  AMICI'S EAST COAST PIZZERIA   
4  100017 2019-08-26  2019           True  AMICI'S EAST COAST PIZZERIA   

            address  
0  PIER 39  K-106-B  
1  PIER 39  K-106-B  
2       475 06th St  
3       475 06th St  
4       475 06th St

In [72]:
ok.grade("q5a");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 3
    Failed: 0
[ooooooooook] 100.0% passed



<br/><br/><br/>

--- 

## Question 5b

Let's look at the 20 businesses with the lowest **median** score.  Order your results by the median score followed by the business id to break ties. The resulting table should look like:


*Hint: You may find the `as_index` argument important*

<table border="1" class="dataframe">  <thead>    <tr style="text-align: right;">      <th></th>      <th>bid</th>      <th>name</th>      <th>median score</th>    </tr>  </thead>  <tbody>    <tr>      <th>3876</th>      <td>84590</td>      <td>Chaat Corner</td>      <td>54.0</td>    </tr>    <tr>      <th>4564</th>      <td>90622</td>      <td>Taqueria Lolita</td>      <td>57.0</td>    </tr>    <tr>      <th>4990</th>      <td>94351</td>      <td>VBowls LLC</td>      <td>58.0</td>    </tr>    <tr>      <th>2719</th>      <td>69282</td>      <td>New Jumbo Seafood Restaurant</td>      <td>60.5</td>    </tr>    <tr>      <th>222</th>      <td>1154</td>      <td>SUNFLOWER RESTAURANT</td>      <td>63.5</td>    </tr>    <tr>      <th>1991</th>      <td>39776</td>      <td>Duc Loi Supermarket</td>      <td>64.0</td>    </tr>    <tr>      <th>2734</th>      <td>69397</td>      <td>Minna SF Group LLC</td>      <td>64.0</td>    </tr>    <tr>      <th>3291</th>      <td>78328</td>      <td>Golden Wok</td>      <td>64.0</td>    </tr>    <tr>      <th>4870</th>      <td>93150</td>      <td>Chez Beesen</td>      <td>64.0</td>    </tr>    <tr>      <th>4911</th>      <td>93502</td>      <td>Smoky Man</td>      <td>64.0</td>    </tr>    <tr>      <th>5510</th>      <td>98995</td>      <td>Vallarta's Taco Bar</td>      <td>64.0</td>    </tr>    <tr>      <th>1457</th>      <td>10877</td>      <td>CHINA FIRST INC.</td>      <td>64.5</td>    </tr>    <tr>      <th>2890</th>      <td>71310</td>      <td>Golden King Vietnamese Restaurant</td>      <td>64.5</td>    </tr>    <tr>      <th>4352</th>      <td>89070</td>      <td>Lafayette Coffee Shop</td>      <td>64.5</td>    </tr>    <tr>      <th>505</th>      <td>2542</td>      <td>PETER D'S RESTAURANT</td>      <td>65.0</td>    </tr>    <tr>      <th>2874</th>      <td>71008</td>      <td>House of Pancakes</td>      <td>65.0</td>    </tr>    <tr>      <th>818</th>      <td>3862</td>      <td>IMPERIAL GARDEN SEAFOOD RESTAURANT</td>      <td>66.0</td>    </tr>    <tr>      <th>2141</th>      <td>61427</td>      <td>Nick's Foods</td>      <td>66.0</td>    </tr>    <tr>      <th>2954</th>      <td>72176</td>      <td>Wolfes Lunch</td>      <td>66.0</td>    </tr>    <tr>      <th>4367</th>      <td>89141</td>      <td>Cha Cha Cha on Mission</td>      <td>66.5</td>    </tr>  </tbody></table>


<!--
BEGIN QUESTION
name: q5b
points: 3
-->

In [73]:
twenty_lowest_scoring=ins_named[['bid','name','score']]
twenty_lowest_scoring=twenty_lowest_scoring[twenty_lowest_scoring['score']>0].groupby(['bid','name'], as_index=False).median().sort_values(['score','bid']).head(20).rename(columns={'score':'median score'})
twenty_lowest_scoring

bid                                name  median score
3876  84590                        Chaat Corner          54.0
4564  90622                     Taqueria Lolita          57.0
4990  94351                          VBowls LLC          58.0
2719  69282        New Jumbo Seafood Restaurant          60.5
222    1154                SUNFLOWER RESTAURANT          63.5
1991  39776                 Duc Loi Supermarket          64.0
2734  69397                  Minna SF Group LLC          64.0
3291  78328                          Golden Wok          64.0
4870  93150                         Chez Beesen          64.0
4911  93502                           Smoky Man          64.0
5510  98995                 Vallarta's Taco Bar          64.0
1457  10877                    CHINA FIRST INC.          64.5
2890  71310   Golden King Vietnamese Restaurant          64.5
4352  89070               Lafayette Coffee Shop          64.5
505    2542                PETER D'S RESTAURANT          65.0
2874  71008                   House of Pancakes          65.0
818    3862  IMPERIAL GARDEN SEAFOOD RESTAURANT          66.0
2141  61427                        Nick's Foods          66.0
2954  72176                        Wolfes Lunch          66.0
4367  89141              Cha Cha Cha on Mission          66.5

In [74]:
ok.grade("q5b");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 2
    Failed: 0
[ooooooooook] 100.0% passed



<br/><br/><br/>

--- 

## Question 5c

Let's now examine the descriptions of violations for inspections with `score > 0` and `score < 65`.  Construct a **Series** indexed by the `description` of the violation from the `vio` table with the value being the number of times that violation occured for inspections with the above score range.  Sort the results in descending order of the count.

The first few entries should look like:

```
Unclean or unsanitary food contact surfaces                                  43
High risk food holding temperature                                           42
Unclean or degraded floors walls or ceilings                                 40
Unapproved or unmaintained equipment or utensils                             39
```
You will need to use `merge` twice.

<!--
BEGIN QUESTION
name: q5c
points: 2
-->

In [75]:
right=pd.merge(vio,ins2vio,on='vid')
main=pd.merge(right,ins[['iid','score']], on='iid')
main=main[main['score']>0]
main=main[main['score']<65]
low_score_violations = pd.Series(main.description.value_counts(),index=main.description.unique()).sort_values(ascending=False)
low_score_violations.head(10)

Unclean or unsanitary food contact surfaces                                  43
High risk food holding temperature                                           42
Unclean or degraded floors walls or ceilings                                 40
Unapproved or unmaintained equipment or utensils                             39
High risk vermin infestation                                                 37
Foods not protected from contamination                                       37
Inadequate food safety knowledge or lack of certified food safety manager    35
Inadequate and inaccessible handwashing facilities                           35
Improper thawing methods                                                     30
Unclean hands or improper use of gloves                                      27
Name: description, dtype: int64

In [76]:
ok.grade("q5c");

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Running tests

---------------------------------------------------------------------
Test summary
    Passed: 3
    Failed: 0
[ooooooooook] 100.0% passed



<br/><br/><br/>

---

<br/><br/><br/>

# 6: Compute Something Interesting

Play with the data and try to compute something interesting about the data. Please try to use at least one of groupby, pivot, or merge (or all of the above).  

Please show your work in the cell below and describe in words what you found in the same cell. This question will be graded leniently but good solutions may be used to create future homework problems.

**Please have both your code and your explanation in the same one cell below. Any work in any other cell will not be graded.**


<!--
BEGIN QUESTION
name: q6
points: 3
manual: True
-->

<!-- EXPORT TO PDF -->

In [77]:
right=pd.merge(ins[['score','bid']],bus[['bid','name','phone_number','latitude','longitude']],on='bid')
right=right[right['score']>0].groupby(['bid','name'], as_index=False).sum().sort_values(['score','bid'],ascending=False)
right.head(90)
num_phone=pd.Series(right['phone_number'].value_counts(),index=right['phone_number'].unique()).sort_values(ascending=False)
right['phone']=~right['phone_number'].apply(str).str.len()==11
count_t=(right['phone']==True).count()
count_f=(right['phone']==False).count()
right.head(60)


#Form the data I collected, we can estblishe some intresting facts:
#    1.The top scored buildings are school buidlings(first at least 90th rank by score is all school)
#    2.High ranking building have all the latitude and longitude for most of the time
#    3.Most of school building who doens't have correct phone number start with '-' and have 5 digit
#        (ex: -19998,-29997,-39996,-49995,-59994, etc)
#    4.Out of 5724 building, only 5724 have the correct phone number
#        (which mena half of the building have phoen while other half doesn't)
#    5.Some of the buildign have the same latitude, phone number, and logntitude, which is false result.
#
#








bid                                     name  score  phone_number  \
1151   5855                SOTA (SCHOOL OF THE ARTS)    778  113245565600   
1205   5953                 ST PETER'S PARISH SCHOOL    696        -69993   
1107   5801                  FLYNN ELEMENTARY SCHOOL    692        -69993   
1126   5823                WEBSTER ELEMENTARY SCHOOL    678   99089870509   
1137   5838            HARVEY MILK ELEMENTARY SCHOOL    676   99086723932   
1150   5854                 Gateway High/Kip Schools    661   99090205200   
3033  74955                            Darioush Inc.    655        -69993   
1164   5871                 CHAVEZ ELEMENTARY SCHOOL    600   84934174590   
1155   5860                              June Jordan    600   84932787282   
1110   5804         JUNIPERO SERRA ELEMENTARY SCHOOL    600   84934174110   
1183   5898                 MONROE ELEMENTARY SCHOOL    598   84932788416   
1180   5894      SAN FRANCISCO COMMUNITY ELEM.SCHOOL    598   84932788434   
1157   5864          VISITACION VALLEY MIDDLE SCHOOL    598        -59994   
1188   5907             LILIENTHAL ELEMENTARY SCHOOL    596   84934461096   
1175   5888                 Bryant Elementary School    596   84934175250   
1148   5852                 CARVER ELEMENTARY SCHOOL    596   84931989240   
1145   5849        NEW TRADITIONS ALTERNATIVE SCHOOL    596   84934550940   
1141   5843                     HOOVER MIDDLE SCHOOL    596   84934516698   
1132   5831         FRANK MCCOPPIN ELEMENTARY SCHOOL    596        -59994   
1123   5819               MIRALOMA ELEMENTARY SCHOOL    596   84932788404   
3116  76223  AT&T PARK - (CART 6) Espresso & Dessert    594        -59994   
1485  13374                    Rooftop Middle School    594   84933160542   
1178   5892                      MISSION HIGH SCHOOL    594        -59994   
1168   5877      CLAIRE LILIENTHAL SCHOOL-MADISON CA    594   84934551618   
1144   5848     Ida B Wells Continuation High School    594   84931477890   
1131   5830         Theodore Roosevelt Middle School    594   84934550676   
1130   5829              LAFAYETTE ELEMENTARY SCHOOL    594   84934550898   
1127   5825              GUADALUPE ELEMENTARY SCHOOL    594   84932788308   
1125   5821                MOSCONE ELEMENTARY SCHOOL    594   84934174416   
1120   5816            WEST PORTAL ELEMENTARY SCHOOL    594   84934517076   
1153   5858                   DREW ELEMENTARY SCHOOL    593   84931989156   
1165   5872                 REVERE ELEMENTARY SCHOOL    592   84934173936   
1162   5869      VISITACION VALLEY ELEMENTARY SCHOOL    592   84932788776   
1160   5867              MALCOLM X ELEMENTARY SCHOOL    592   84934175700   
1147   5851                           ROOFTOP SCHOOL    592   84934174152   
1122   5818            MISSION ED. CENTER ELEMENTARY    591   84934171878   
1139   5841              JEFFERSON ELEMENTARY SCHOOL    590   84934516926   
1134   5834                   Presidio Middle School    590   84934550610   
1113   5808                       BALBOA HIGH SCHOOL    590   84932784540   
1190   5912      ARGONNE ALTERNATIVE ELEMENTARY SCHO    589   84934550760   
1179   5893              CLEVELAND ELEMENTARY SCHOOL    589   84932788254   
2870  70961        Our Lady of the Visitacion School    588        -59994   
1193   5919                 SAN FRANCISCO DAY SCHOOL    588   84935594532   
1158   5865                 TAYLOR ELEMENTARY SCHOOL    588   84931989180   
1734  33413             Creative Arts Charter school    587        -59994   
1185   5902                      Galileo High School    587        -59994   
1161   5868             LONGFELLOW ELEMENTARY SCHOOL    587   84932788380   
1181   5895                     Downtown High School    586   84934175196   
1108   5802                       LOWELL HIGH SCHOOL    586   84934516488   
1105   5797                  SLOAT ELEMENTARY SCHOOL    586        -59994   
1172   5884           Gordon J Lau Elementary School    585        -59994   
1156

In [78]:
#THIS CELL AND ANY CELLS ADDED BELOW WILL NOT BE GRADED

## Congratulations! You have finished Part 1 of Project 1! 

In our analysis of the business data, we found that there are some errors with the ZIP codes. As a result, we made the records with ZIP codes outside of San Francisco have a `postal5` value of `None` and shortened 9-digit zip codes to 5-digit ones. In practice, we could take the time to look up the restaurant address online and fix some of the zip code issues.

In our analysis of the inspection data, we investigated the relationship between the year and type of inspection, and we figured out that only `Routine - Unscheduled` inspections have inspection scores.
Finally, we joined the business and inspection data to identify restaurants with the worst ratings and the lowest median scores. 

# Submit
Make sure you have run all cells in your notebook in order before running the cell below, so that all images/graphs appear in the output.
**Please save before submitting!**

<!-- EXPECT 2 EXPORTED QUESTIONS -->

In [79]:
# Save your notebook first, then run this cell to submit.
import jassign.to_pdf
jassign.to_pdf.generate_pdf('proj1a.ipynb', 'proj1a.pdf')
ok.submit()

Generating PDF...
Saved proj1a.pdf


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Saving notebook... Saved 'proj1a.ipynb'.
Submit... 100% complete
Submission successful for user: ellierobin54@berkeley.edu
URL: https://okpy.org/cal/data100/sp20/proj1a/submissions/DqgB6k

